clients/roscpp/src/libros/init.cpp

```diff
@@ -285,6 +285,33 @@ void internalCallbackQueueThreadFunc()
   }
 }
 
+struct InternalQueueJoiningThread
+{
+  InternalQueueJoiningThread()
+  {
+      g_internal_queue_thread = boost::thread(internalCallbackQueueThreadFunc);
+  }
+
+  ~InternalQueueJoiningThread()
+  {
+    if (g_internal_queue_thread.joinable())
+    {
+      g_shutting_down = true;
```

Comment:
Why do you need to set this flag? This essentially prohibits `shutdown()` from being executed.

Reply:
> Why do you need to set this flag?

My motivation was to signal the internalCallbackQueueThreadFunc thread to exit its main loop.

```
void internalCallbackQueueThreadFunc()
{
  disableAllSignalsInThisThread();

  CallbackQueuePtr queue = getInternalCallbackQueue();

  while (!g_shutting_down)
  {
    queue->callAvailable(WallDuration(0.1));
  }
}
```

> This essentially prohibits shutdown() from being executed.

Right, thanks for catching this.

Let me look again at how to properly do this.